pkg/kubelet/cm/devicemanager/pod_devices.go

```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
 			annotationsMap[k] = v
 			opts.Annotations = append(opts.Annotations, kubecontainer.Annotation{Name: k, Value: v})
 		}
+
+		// Updates for CDI devices. This also updates the annotations.
+		cdiDevices, cdiAnnotations := getCDIDeviceInfo(
+			podUID, resourceName,
+			resp.CdiDevices,
+			annotationsMap,
+			cdiDevicesMap,
+		)
+		opts.CDIDevices = append(opts.CDIDevices, cdiDevices...)
+		opts.Annotations = append(opts.Annotations, cdiAnnotations...)
 	}
 	return opts
 }
 
+// getCDIDeviceInfo returns CDI devices and annotations for a given pod.
+func getCDIDeviceInfo(podUID, resourceName string, responseCDIDevices []*pluginapi.CDIDevice, annotationsMap map[string]string, cdiDevicesMap map[string]struct{}) ([]kubecontainer.CDIDevice, []kubecontainer.Annotation) {
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DevicePluginCDIDevices) {
+		return nil, nil
+	}
+
+	var cdiDeviceNames []string
+	for _, cdiDevice := range responseCDIDevices {
+		if _, ok := cdiDevicesMap[cdiDevice.Name]; ok {
+			klog.V(4).InfoS("Skip existing CDI Device", "name", cdiDevice.Name)
+			continue
+		}
+		klog.V(4).InfoS("Add CDI device", "name", cdiDevice.Name)
+		cdiDevicesMap[cdiDevice.Name] = struct{}{}
+
+		cdiDeviceNames = append(cdiDeviceNames, cdiDevice.Name)
```

Comment:
*IFF* `cdiDevicesMap` is actually behaving like a set, and if we switch to `apimachinery/util/set` I think we will be able to simplify this code. Worth a shot.

Reply:
Using a `Set` here would only solve for local (per-resource) uniqueness and not global uniqueness. Maybe that is an indication that using a set is not what we want and we should store the resource name in our `cdiDevicesMap` to provide logs as to which resource is also requesting a specific CDI device.